sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/GrowableOffsetRangeTracker.java

```diff
@@ -0,0 +1,103 @@
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms.splittabledofn;
+
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkNotNull;
+
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.io.range.OffsetRange;
+
+/**
+ * A special {@link OffsetRangeTracker} for tracking a growable offset range. The Long.MAX_VALUE is
+ * used as end range to indicate the possibility of infinity.
+ *
+ * <p>A offset range is considered as growable when the end offset could grow(or change) during
+ * execution time(e.g., Kafka backlog, appended file).
+ */
+@Experimental(Kind.SPLITTABLE_DO_FN)
+public class GrowableOffsetRangeTracker extends OffsetRangeTracker {
+  /**
+   * An interface that should be implemented to fetch estimated end offset of range.
+   *
+   * <p>{@code estimateRangeEnd} is called to give te end offset when {@code trySplit} or {@code
+   * getProgress} is invoked. The end offset is exclusive for the range. It's not necessary to
+   * increase monotonically but it's only taken into computation when it's larger than the current
+   * position. When returning Long.MAX_VALUE as estimate, it means the largest possible position for
+   * the range is Long.MAX_VALUE - 1. Having a good estimate is important for providing a good
+   * signal of progress and splitting at a proper position.
+   */
+  public interface OffsetPoller {
+    long estimateRangeEnd();
+  }
+
+  private final OffsetPoller poller;
+
+  public GrowableOffsetRangeTracker(long start, OffsetPoller offsetPoller) {
+    super(new OffsetRange(start, Long.MAX_VALUE));
+    this.poller = checkNotNull(offsetPoller);
+  }
+
+  @Override
+  public SplitResult<OffsetRange> trySplit(double fractionOfRemainder) {
+    // If current tracking range is no longer growable, split it as a normal range.
+    if (range.getTo() != Long.MAX_VALUE || range.getTo() == range.getFrom()) {
+      return super.trySplit(fractionOfRemainder);
+    }
+    long cur = (lastAttemptedOffset == null) ? range.getFrom() - 1 : lastAttemptedOffset;
+
+    // If current range has been done, there is no more space to split.
+    if (cur == Long.MAX_VALUE) {
+      return null;
+    }
+    // Fetch the estimated end offset. If the estimated end is smaller than the next offset, use
+    // the next offset as end.
+    long estimateLatestOffset = Long.max(poller.estimateRangeEnd(), cur + 1);
+    long splitPos =
+        cur
+            + Math.max(
+                1L,
+                (Double.valueOf((estimateLatestOffset - cur) * fractionOfRemainder)).longValue());
+
+    if (splitPos > estimateLatestOffset) {
+      return null;
+    }
+    OffsetRange res = new OffsetRange(splitPos, range.getTo());
+    this.range = new OffsetRange(range.getFrom(), splitPos);
+    return SplitResult.of(range, res);
```

Comment:
Why not construct a new OffsetRangeTracker over `[cur, estimateLatestOffset)` and call trySplit on it?

Reply:
There is a slight difference around comparison of `splitPos` and `end`.  `GrowableOffsetTracker` allows split when `splitPos <= end` but `OffsetRangeTracker` allows split when `splitPos < end`. The purpose of `OffsetRangeTracker` is to reduce resuming from an empty range. The purpose of `GrowableOffsetRangeTracker` is to handle split at `cur + 1`.

Reply:
btw we will also loss `lastAttemptedOffset` by creating a new tracker.